src/api.ts

```diff
@@ -0,0 +1,559 @@
+   * is the happy path and is modeled explicitly so that the return type of the
+   * integration can be inferred. This will be compiled to a VTL template.
+   *
+   * At present the function body must be a single statement returning an object
+   * literal. The supported syntax will be expanded in the future.
+   */
+  response: ResponseTransformerFunction<IntegrationResponse, MethodResponse>;
+  /**
+   * Map of status codes to a function defining the  response to return. This is used
+   * to configure the failure path method responses, for e.g. when an integration fails.
+   */
+  errors: { [statusCode: number]: () => any };
+}
+
+/**
+ * An AWS API Gateway integration lets you integrate an API with an AWS service
+ * supported by Functionless. The request is transformed via VTL and sent to the
+ * service via API call, and the response is transformed via VTL and returned in
+ * the response.
+ *
+ * Only `application/json` is supported.
+ *
+ * TODO: provide example usage after api is stabilized
+ *
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-api-integration-types.html
+ */
+export class AwsApiIntegration<
+  Props extends AwsApiIntegrationProps<any, any, any>
+> extends BaseApiIntegration {
+  private readonly request: FunctionDecl;
+  private readonly response: FunctionDecl;
+  private readonly errors: { [statusCode: number]: FunctionDecl };
+
+  public constructor(props: Props) {
+    super();
+    this.request = validateFunctionDecl(props.request);
+    this.response = validateFunctionDecl(props.response);
+    this.errors = Object.fromEntries(
+      Object.entries(props.errors).map(([k, v]) => [k, validateFunctionDecl(v)])
+    );
+  }
+
+  public addMethod(
+    httpMethod: HttpMethod,
+    resource: aws_apigateway.Resource
+  ): void {
+    const [requestTemplate, integration] = toVTL(this.request, "request");
+    const [responseTemplate] = toVTL(this.response, "response");
+
+    const errorResponses: aws_apigateway.IntegrationResponse[] = Object.entries(
+      this.errors
+    ).map(([statusCode, fn]) => {
+      const [template] = toVTL(fn, "response");
+      return {
+        statusCode: statusCode,
+        selectionPattern: `^${statusCode}$`,
+        responseTemplates: {
+          "application/json": template,
+        },
+      };
+    });
+
+    const integrationResponses: aws_apigateway.IntegrationResponse[] = [
+      {
+        statusCode: "200",
+        responseTemplates: {
+          "application/json": responseTemplate,
+        },
+      },
+      ...errorResponses,
+    ];
+
+    // TODO: resource is not the right scope, prevents adding 2 methods to the resource
+    // because of the IAM roles created
+    // should `this` be a Method?
+    const apiGwIntegration = integration!.apiGWVtl.makeIntegration(
+      resource,
+      requestTemplate,
+      integrationResponses
+    );
+
+    const methodResponses = [
+      { statusCode: "200" },
```

Comment:
whatever approach we take for above can be done for the success too. 99% of the time people just mean 200 though so it needs to be unobtrusive.

sometimes people want multiple 2XX responses too. But I don't think we need to break our backs supporting all edge cases yet